qiskit/pulse/schedule.py

```diff
@@ -1066,15 +1267,40 @@ def append(
             )
 
         if not inplace:
-            ret_block = copy.deepcopy(self)
-            ret_block._name = name or self.name
-            ret_block.append(block, inplace=True)
-            return ret_block
-        else:
-            self._blocks.append(block)
-            self._parameter_manager.update_parameter_table(block)
+            schedule = copy.deepcopy(self)
+            schedule._name = name or self.name
+            schedule.append(block, inplace=True)
+            return schedule
+
+        if isinstance(block, Reference):
+            self.references[block.ref_keys] = None
+
+        elif isinstance(block, ScheduleBlock):
+            block = copy.deepcopy(block)
+            # Expose subroutines to the current main scope.
+            # Note that this 'block' is not called.
+            # The block is just directly appended to the current scope.
+            if block.is_referenced():
+                if block._parent is not None:
+                    # This is an edge case:
+                    # If this is not a parent, block.references points to the parent's reference
+                    # where subroutine not referred within the 'block' may exist.
+                    # Move only references existing in the 'block'.
+                    # See 'test.python.pulse.test_reference.TestReference.test_appending_child_block'
+                    for ref in _get_references(block._blocks):
+                        self.references[ref.ref_keys] = block.references[ref.ref_keys]
+                else:
+                    # This operation is much faster because this is dict update.
+                    # Reference manager of appended block is cleared because of data reduction.
+                    self.references.update(block._reference_manager)
+                    block._reference_manager.clear()
+            # Now switch the parent because block is appended to self.
+            block._parent = self
```

Comment:
No, this is due to nesting feature of schedule blocks. For example,

```python
with build() as root:
    with align_left():
        with align_right():
            reference("sub")
```
In this example, we should be able to assign subroutine with
```python
root.assign_references({("sub", ): subroutine})
```
because every reference in the same builder context should be exposed to `root`, i.e.  nesting a block doesn't create reference. However, because of current implementation bottleneck, ref manager is attached to each block, and thus, without parent referencing mechanism, we must write this awkward code:
```python
root.blocks[0].blocks[0].assign_references({("sub", ): subroutine})
```
The parent reference allows us to manage references defined in all nested blocks in the most outer block.

Reply:
Okay, that makes sense. Maybe a comment on `_parent` in `__init__` would be good. One thing I wonder about is substituted references. If `_parent` is only ever used for looking up the reference manager, there should be no problem since `.blocks` substitutes in the references and so references should not need to be considered further. When you call `.blocks` you will get a list of blocks with `_parent` pointing at the parent, except for the assigned references which will be `ScheduleBlock`s with `_parent == None`. So I could see potential confusion in the future if `_parent` were used in some other way.

Reply:
Thanks, Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/620e67d62b584537cbd05c741bbeea0dab80fe6a